File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
import java.util.*;

class Solution {
    public int[] productExceptSelf(int[] nums) {
        // O(n) without division operation
        // 5 <- 1 
        // bcd acd abd abc
        //1 [1, 2, 6, 24] 1 
        //  a  ab abc abcd
        //1 [24, 24, 12, 4] 1
        //  dcba dcb dc d
        
        // initialize
        // 4
        int len = nums.length;
        int[] prefix = new int[len + 2];
        int[] suffix = new int[len + 2];
        prefix[0] = 1;
        suffix[0] = 1;
        prefix[len + 1] = 1;
        suffix[len + 1] = 1;

        for (int i = 0; i < len; i++) {
            prefix[i + 1] = nums[i] * prefix[i];
        }

        for (int i = len - 1; i >= 0; i--) {
            suffix[i + 1] = nums[i] * suffix[i + 2];
        }

        // System.out.println(Arrays.toString(prefix));
        // System.out.println(Arrays.toString(suffix));

        //[1, 1, 2, 6, 24, 1] prefix
        // [1, 24, 24, 12, 4, 1] suffix

        int[] result = new int[len];
        for (int i = 0; i < len; i++) {
            result[i] = prefix[i] * suffix[i + 2];
        }
        
        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [1] + [0]*n + [1]
        suffix = [1] + [0]*n + [1]

        for i in range(n):
            prefix[i + 1] = prefix[i] * nums[i]
        
        for i in range(n - 1, -1, -1):
            suffix[i + 1] = suffix[i + 2] * nums[i]
        
        res = []
        for i in range(n):
            product = prefix[i] * suffix[i + 2]
            res.append(product)
        
        return res